File: numass-main/src/main/java/inr/numass/actions/AnalyzeDataAction.java
package inr.numass.actions;

import hep.dataforge.actions.OneToOneAction;
import hep.dataforge.context.Context;
import hep.dataforge.description.TypedActionDef;
import hep.dataforge.description.ValueDef;
import hep.dataforge.meta.Laminate;
import hep.dataforge.tables.Table;
import inr.numass.data.analyzers.SmartAnalyzer;
import inr.numass.data.api.NumassAnalyzer;
import inr.numass.data.api.NumassSet;
import inr.numass.utils.NumassUtils;

import static hep.dataforge.values.ValueType.NUMBER;
import static hep.dataforge.values.ValueType.STRING;

/**
 * The action performs the readout of data and collection of count rate into a table
 * Created by darksnake on 11.07.2017.
 */
@TypedActionDef(name = "numass.analyze", inputType = NumassSet.class, outputType = Table.class)
@ValueDef(name = "window.lo", type = {NUMBER, STRING}, def = "0", info = "Lower bound for window")
@ValueDef(name = "window.up", type = {NUMBER, STRING}, def = "10000", info = "Upper bound for window")
public class AnalyzeDataAction extends OneToOneAction<NumassSet, Table> {
    @Override
    protected Table execute(Context context, String name, NumassSet input, Laminate inputMeta) {
        //TODO add processor here
        NumassAnalyzer analyzer = new SmartAnalyzer();
        Table res = analyzer.analyzeSet(input, inputMeta);
        output(context, name, stream -> NumassUtils.writeSomething(stream, inputMeta, res));
        return res;
    }
}

File: numass-main/src/main/kotlin/inr/numass/actions/AnalyzeDataAction.kt
package inr.numass.actions

import hep.dataforge.actions.OneToOneAction
import hep.dataforge.context.Context
import hep.dataforge.description.TypedActionDef
import hep.dataforge.description.ValueDef
import hep.dataforge.description.ValueDefs
import hep.dataforge.meta.Laminate
import hep.dataforge.tables.Table
import hep.dataforge.values.ValueType.NUMBER
import hep.dataforge.values.ValueType.STRING
import inr.numass.NumassUtils
import inr.numass.data.analyzers.NumassAnalyzer
import inr.numass.data.api.NumassSet
import inr.numass.data.analyzers.SmartAnalyzer

/**
 * The action performs the readout of data and collection of count rate into a table
 * Created by darksnake on 11.07.2017.
 */
@TypedActionDef(name = "numass.analyze", inputType = NumassSet::class, outputType = Table::class)
@ValueDefs(
        ValueDef(key = "window.lo", type = arrayOf(NUMBER, STRING), def = "0", info = "Lower bound for window"),
        ValueDef(key = "window.up", type = arrayOf(NUMBER, STRING), def = "10000", info = "Upper bound for window")
)
object AnalyzeDataAction : OneToOneAction<NumassSet, Table>("numass.analyze", NumassSet::class.java, Table::class.java) {
    override fun execute(context: Context, name: String, input: NumassSet, inputMeta: Laminate): Table {
        //TODO add processor here
        val analyzer: NumassAnalyzer = SmartAnalyzer()
        val res = analyzer.analyzeSet(input, inputMeta)

        render(context, name, NumassUtils.wrap(res, inputMeta))
//        output(context, name) { stream -> NumassUtils.write(stream, inputMeta, res) }
        return res
    }
}
